cmd/crossplane/core/core.go

```diff
@@ -238,20 +245,14 @@ func (c *startCommand) Run(s *runtime.Scheme, log logging.Logger) error { //noli
 		return errors.Wrap(err, "Cannot add packages controllers to manager")
 	}
 
-	if c.WebhookTLSCertDir != "" {
```

Comment:
Does this mean we are going to deploy webhooks always now? (I have to check the code better, just wanted to drop a comment before forgetting and before this got merged)

Reply:
Yeah this is something I want to understand better too. I _think_ before this change:

* A webhook server was always started and listening (by controller runtime).
* If `c.WebhookTLSCertDir` was supplied, we'd tweak its config and register our webhooks.
* We'd only actually expose the webhook server if `webhooks.enabled` was true in the Helm chart

My thinking with this change was it might be fine to just register the webhooks whether they were going to be exposed/used or not.

Reply:
Hrm, no this now breaks. I get the following when I set `webhooks.enabled=false`:

```
crossplane: error: core.startCommand.Run(): Cannot start controller manager: open /tmp/k8s-webhook-server/serving-certs/tls.crt: no such file or directory
```

Reply:
yes, that was what I was expecting, due to how webhooks are handled we also hit this issue: https://github.com/crossplane/crossplane/issues/4095
I don't think we should address it here, but could be worth keeping it in mind and if it becomes to complex we could think of alternatives, e.g. moving validatingwebhookconfiguration creation to the helm chart and leave in the controller only the ca injection

Reply:
I tweaked things a bit (see new commits) but it's still not passing the E2E suite in https://github.com/crossplane/crossplane/pull/4187.

Reply:
Looks like a lot of breaking changes related to webhooks in controller-runtime v0.15. I've added a commit that implements an `admission.CustomValidator`, but I'm still seeing failures when I run the E2E tests. Interestingly the failures are now to do with XRDs. 🤔 

```
2023-06-22T09:10:09Z    ERROR   crossplane      Reconciler error        {"controller": "defined/compositeresourcedefinition.apiextensions.crossplane.io", "controllerGroup": "apiextensions.crossplane.io", "controllerKind": "CompositeResourceDefinition", "CompositeResourceDefinition": {"name":"xnopresources.nop.example.org"}, "namespace": "", "name": "xnopresources.nop.example.org", "reconcileID": "3defa3c9-085d-4b2f-9f77-b88eb7322ec3", "error": "cannot add composite resource finalizer: cannot update object: Internal error occurred: failed calling webhook \"compositeresourcedefinitions.apiextensions.crossplane.io\": failed to call webhook: the server could not find the requested resource"}
```

Reply:
Found it - XRD needed its methods updated to be a valid `admission.Validator`. Pushed a commit.